google/cloud/bigtable/_mutate_rows.py

```diff
@@ -0,0 +1,239 @@
+      - operation_timeout: the timeout to use for the entire operation, in seconds.
+      - per_request_timeout: the timeout to use for each mutate_rows attempt, in seconds.
+          If not specified, the request will run until operation_timeout is reached.
+      - on_terminal_state: If given, this function will be called as soon as a mutation entry
+            reaches a terminal state (success or failure).
+    """
+    mutations_dict: dict[int, RowMutationEntry] = {
+        idx: mut for idx, mut in enumerate(mutation_entries)
+    }
+
+    error_dict: dict[int, list[Exception]] = {idx: [] for idx in mutations_dict.keys()}
+
+    predicate = retries.if_exception_type(
+        core_exceptions.DeadlineExceeded,
+        core_exceptions.ServiceUnavailable,
+        _MutateRowsIncomplete,
+    )
+
+    def on_error_fn(exc):
+        if predicate(exc) and not isinstance(exc, _MutateRowsIncomplete):
+            # add this exception to list for each active mutation
+            for idx in error_dict.keys():
+                error_dict[idx].append(exc)
+            # remove non-idempotent mutations from mutations_dict, so they are not retried
+            for idx, mut in list(mutations_dict.items()):
+                if not mut.is_idempotent():
+                    mutations_dict.pop(idx)
+
+    retry = retries.AsyncRetry(
+        predicate=predicate,
+        on_error=on_error_fn,
+        timeout=operation_timeout,
+        initial=0.01,
+        multiplier=2,
+        maximum=60,
+    )
+    # use generator to lower per-attempt timeout as we approach operation_timeout deadline
+    attempt_timeout_gen = _attempt_timeout_generator(
+        per_request_timeout, operation_timeout
+    )
+    # wrap attempt in retry logic
+    retry_wrapped = retry(_mutate_rows_retryable_attempt)
+    # convert RetryErrors from retry wrapper into DeadlineExceeded errors
+    deadline_wrapped = _convert_retry_deadline(retry_wrapped, operation_timeout)
+    try:
+        # trigger mutate_rows
+        await deadline_wrapped(
+            gapic_client,
+            table,
+            attempt_timeout_gen,
+            mutations_dict,
+            error_dict,
+            predicate,
+            on_terminal_state,
+        )
+    except Exception as exc:
+        # exceptions raised by retryable are added to the list of exceptions for all unfinalized mutations
+        for error_list in error_dict.values():
+            error_list.append(exc)
+    finally:
+        # raise exception detailing incomplete mutations
+        all_errors = []
+        for idx, exc_list in error_dict.items():
+            if len(exc_list) == 0:
+                raise core_exceptions.ClientError(
+                    f"Mutation {idx} failed with no associated errors"
+                )
+            elif len(exc_list) == 1:
+                cause_exc = exc_list[0]
+            else:
+                cause_exc = bt_exceptions.RetryExceptionGroup(exc_list)
```

Comment:
I was following a similar pattern in other rpcs (including mutate_row), where it will return the exception directly if it's not retried, but wrap a group of exceptions if there are multiple failed tries. This logic is an extension of that, but adding an extra grouping layer to represent multiple entries. 

So do you think we should change just this one and have a bit of an inconsistency there, or change all rpcs to always wrap in a RetryExceptionGroup, or leave as-is?

(This may be worth addressing as an issue in the repo, since it may require changes across a couple places)

Reply:
Lets keep as is for now and add a github issue to take another pass on exception structure

Reply:
Sounds good, I opened one here: https://github.com/googleapis/python-bigtable/issues/794